File: journal/program/ts/state/journal.ts
import { Assignable } from '../util/util';
import * as borsh from "borsh";
import { Buffer } from "buffer";


export class JournalMetadata extends Assignable {
    toBuffer() { 
        return Buffer.from(borsh.serialize(JournalMetadataSchema, this)) 
    };
    
    static fromBuffer(buffer: Buffer) {
        return borsh.deserialize(JournalMetadataSchema, JournalMetadata, buffer);
    };
};

export const JournalMetadataSchema = new Map([
    [ JournalMetadata, { 
        kind: 'struct', 
        fields: [ 
            ['nickname', 'string'],
            ['authority', [32]],
            ['entries', 'u32'],
            ['bump', 'u8'],
        ],
    }]
]);

File: journal/journal-program/src/state/journal.rs
use borsh::{ BorshDeserialize, BorshSerialize };
use solana_program::pubkey::Pubkey;


// The journal account's data
//
#[derive(BorshDeserialize, BorshSerialize, Debug)]
pub struct JournalMetadata {
    pub nickname: String,
    pub authority: Pubkey,
    pub entries: u32,
    pub bump: u8,
}

impl JournalMetadata {

    pub const ACCOUNT_SPACE: usize = 8 + 40 + 32 + 4 + 1;
    pub const SEED_PREFIX: &'static str = "journal";

    pub fn new(
        nickname: String,
        authority: Pubkey,
        bump: u8,
    ) -> Self {
        
        JournalMetadata {
            nickname,
            authority,
            entries: 0,
            bump,
        }
    }

    pub fn increment(&mut self) {
        self.entries += 1;
    }
}